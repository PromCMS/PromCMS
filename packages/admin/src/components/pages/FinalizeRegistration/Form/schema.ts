import { object, string, ref } from 'yup';

export const finalizeRegistrationFormSchema = object()
  .shape({
    new_password: string().required('Dont forget to include this value'),
    confirmed_new_password: string()
      .required('Dont forget to include this value')
      .oneOf([ref('new_password'), null], 'Passwords must match'),
    token: string().required('Dont forget to include this value'),
  })
  .noUnknown();
