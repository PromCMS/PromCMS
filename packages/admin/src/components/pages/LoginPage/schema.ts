import { MESSAGES } from '@constants'
import { number, object, string } from 'yup'

export const loginFormSchema = object()
  .shape({
    step: number().required(),
    password: string().when('step', {
      is: 0,
      then: string().required(MESSAGES.PLEASE_ENTER_PASSWORD),
    }),
    email: string()
      .email()
      .when('step', {
        is: 0,
        then: string().required(MESSAGES.PLEASE_ENTER_EMAIL),
      }),
    token: string().when('step', {
      is: 2,
      then: string()
        .required(MESSAGES.PLEASE_ENTER_MFA_TOKEN)
        .length(6, MESSAGES.MFA_TOKEN_SHORT),
    }),
  })
  .noUnknown()
