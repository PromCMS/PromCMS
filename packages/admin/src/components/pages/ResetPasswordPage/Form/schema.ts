import { object, string, ref } from 'yup'

export const initializeResetPasswordFormSchema = object()
  .shape({
    email: string()
      .email('Please enter valid email address')
      .required('Dont forget to include this value'),
  })
  .noUnknown()

export const finalizeResetPasswordFormSchema = object()
  .shape({
    new_password: string().required('Dont forget to include this value'),
    confirmed_new_password: string()
      .required('Dont forget to include this value')
      .oneOf([ref('new_password'), null], 'Passwords must match'),
    token: string().required('Dont forget to include this value'),
  })
  .noUnknown()
