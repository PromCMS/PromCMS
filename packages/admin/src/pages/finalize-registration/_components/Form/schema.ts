import { object, string, ref } from 'yup';

export const finalizeRegistrationFormSchema = object()
  .shape({
    new_password: string().required('This field is required'),
    confirmed_new_password: string()
      .required('This field is required')
      .oneOf([ref('new_password'), null], 'Passwords must match'),
    token: string().required('This field is required'),
  })
  .noUnknown();
