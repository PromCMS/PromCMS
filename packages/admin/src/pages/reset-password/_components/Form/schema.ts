import { MESSAGES } from '@constants';
import { z } from 'zod';

export const initializeResetPasswordFormSchema = z
  .object({
    email: z
      .string({
        required_error: MESSAGES.FIELD_REQUIRED,
      })
      .email('Please enter valid email address'),
  })
  .strict();

export const finalizeResetPasswordFormSchema = z
  .object({
    new_password: z.string({
      required_error: MESSAGES.FIELD_REQUIRED,
    }),
    confirmed_new_password: z.string({
      required_error: MESSAGES.FIELD_REQUIRED,
    }),
    token: z.string({
      required_error: MESSAGES.FIELD_REQUIRED,
    }),
  })
  .strict()
  .superRefine((value, ctx) => {
    if (value.new_password !== value.confirmed_new_password) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: MESSAGES.PASSWORDS_MUST_MATCH,
      });
    }
  });
