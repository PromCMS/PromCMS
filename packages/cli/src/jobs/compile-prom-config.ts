import { formatCodeString } from '@utils';
import fs from 'fs-extra';
import path from 'node:path';

import { GeneratorConfig } from '@prom-cms/schema';

export type CompilePromConfigOptions = {
  config: GeneratorConfig;
  appRoot: string;
};

const recursivePrintObjectToPhp = (obj: object | Map<any, any> | []) => {
  let result = ``;
  const objAsArray = Array.isArray(obj)
    ? obj.map((item) => [null, item])
    : obj instanceof Map
      ? obj
      : Object.entries(obj);

  const items: string[] = [];

  for (const [key, value] of objAsArray) {
    let formattedValue = '';

    switch (typeof value) {
      case 'boolean':
      case 'number':
        formattedValue = String(value);
        break;
      case 'object':
        formattedValue = recursivePrintObjectToPhp(value);
        break;
      default:
        formattedValue = `'${String(value)}'`;
        break;
    }

    if (key) {
      formattedValue = `'${key}' => ${formattedValue}`;
    }

    items.push(formattedValue);
  }

  result = items.join(', \n');

  return `[${result}]`;
};

export const compilePromConfig = async ({
  config,
  appRoot,
}: CompilePromConfigOptions) => {
  const filepath = path.join(appRoot, '.prom-cms', 'compiled', 'config.php');

  let contents = `<?php\n
/**
 * This file is generated by PromCMS, do not edit this file as changes made to this file will be overriden in the next model sync. 
 * Updates should be made to ../UserRole.php as that is not overriden.
 */

return [\n`;

  for (const [key, value] of Object.entries(config)) {
    contents += `"${key}" => ${recursivePrintObjectToPhp(value)},`;
    contents += '\n';
  }

  contents += '];';

  await fs.ensureDir(path.dirname(filepath));
  await fs.writeFile(
    filepath,
    await formatCodeString(contents, path.basename(filepath))
  );
};
