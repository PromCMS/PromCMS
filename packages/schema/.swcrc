{
  "$schema": "http://json.schemastore.org/swcrc",
  "module": {
    "type": "es6"
  },
  "jsc": {
    "target": "esnext",
    "baseUrl": "./src",
    "parser": {
      "syntax": "typescript",
      "tsx": true
    },
    "transform": {
      "react": {
        "runtime": "automatic"
      }
    }
  }
}