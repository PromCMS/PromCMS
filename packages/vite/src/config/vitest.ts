import reactPlugin from '@vitejs/plugin-react';
import tsconfigPaths from 'vite-tsconfig-paths';
import { defineConfig } from 'vitest/config';

export default defineConfig({
  plugins: [
    tsconfigPaths(),
    // @ts-expect-error
    reactPlugin(),
  ],
  test: {},
});
