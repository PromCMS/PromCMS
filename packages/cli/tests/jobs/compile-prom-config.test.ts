import { MONOREPO_ROOT } from '@constants';
import fs from 'fs-extra';
import path from 'path';
import { afterAll, beforeEach, describe, expect, it } from 'vitest';
import { z } from 'zod';

import { databaseConfigSchema } from '@prom-cms/schema';

import { compilePromConfig } from '../../src/jobs/compile-prom-config.js';

const TEST_FOLDER_PATH = path.join(
  MONOREPO_ROOT,
  'node_modules',
  '.prom-cms',
  'test-app'
);
describe('jobs', () => {
  beforeEach(async () => {
    await fs.remove(TEST_FOLDER_PATH);
    await fs.ensureDir(TEST_FOLDER_PATH);
  });

  afterAll(async () => {
    await fs.remove(TEST_FOLDER_PATH);
  });

  describe('compilePromConfig', () => {
    it('should run correctly and generate files', async () => {
      const root = path.join(TEST_FOLDER_PATH, 'modules', 'test');
      await fs.ensureDir(root);

      await compilePromConfig({
        config: {
          project: { name: 'Testing Project', url: '', languages: ['en'] },
          database: databaseConfigSchema.parse({
            connections: [
              {
                name: 'default-connection',
                uri: `pdo-sqlite:///${path.join(
                  '.database',
                  'application.sqlite'
                )}`,
              },
            ],
            models: [
              {
                tableName: 'first',
                title: 'First',
                admin: { isHidden: true, icon: 'Article' },
                columns: [
                  { name: 'col', type: 'string', title: 'New' },
                  {
                    name: 'bola',
                    type: 'boolean',
                    localized: true,
                    title: 'Boolean Column',
                    defaultValue: false,
                  },
                ],
              },
              {
                title: 'Second',
                admin: { icon: 'Article' },
                tableName: 'second_table_name',
                api: { enabled: false },
                columns: [{ name: 'col', type: 'string', title: 'Newo' }],
              },
            ],
          } satisfies z.input<typeof databaseConfigSchema>),
        },
        appRoot: TEST_FOLDER_PATH,
      });

      const expectedFiles = [
        path.join(TEST_FOLDER_PATH, '.prom-cms/parsed/config.php'),
      ];

      for (const filePath of expectedFiles) {
        expect(
          await fs.readFile(filePath, {
            encoding: 'utf8',
          })
        ).toMatchSnapshot();
      }
    });
  });
});
