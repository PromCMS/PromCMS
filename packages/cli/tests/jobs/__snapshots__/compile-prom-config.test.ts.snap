// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`jobs > compilePromConfig > should run correctly and generate files 1`] = `
"<?php

/**
 * This file is generated by PromCMS, do not edit this file as changes made to this file will be overriden in the next model sync.
 */

return [
  'project' => [
    'name' => 'Testing Project',
    'url' => '',
    'languages' => ['en'],
  ],
  'database' => [
    'connections' => [
      [
        'name' => 'default-connection',
        'uri' => 'pdo-sqlite:///.database/application.sqlite',
      ],
    ],
    'models' => [
      [
        'tableName' => 'first',
        'title' => 'First',
        'ignoreSeeding' => false,
        'timestamp' => true,
        'admin' => ['isHidden' => true, 'icon' => 'Article'],
        'api' => ['enabled' => true],
        'columns' => [
          [
            'title' => 'New',
            'name' => 'col',
            'required' => true,
            'unique' => false,
            'localized' => false,
            'admin' => [
              'isHidden' => false,
              'editor' => ['placement' => 'main', 'width' => 12],
              'fieldType' => 'normal',
            ],
            'readonly' => false,
            'hide' => false,
            'type' => 'string',
            'primaryString' => false,
          ],
          [
            'title' => 'Boolean Column',
            'name' => 'bola',
            'required' => true,
            'unique' => false,
            'localized' => true,
            'admin' => [
              'isHidden' => false,
              'editor' => ['placement' => 'main', 'width' => 12],
            ],
            'readonly' => false,
            'hide' => false,
            'type' => 'boolean',
            'defaultValue' => false,
          ],
        ],
        'softDelete' => false,
        'draftable' => false,
        'sorting' => false,
        'sharable' => true,
        'ownable' => true,
      ],
      [
        'tableName' => 'second_table_name',
        'title' => 'Second',
        'ignoreSeeding' => false,
        'timestamp' => true,
        'admin' => ['isHidden' => false, 'icon' => 'Article'],
        'api' => ['enabled' => false],
        'columns' => [
          [
            'title' => 'Newo',
            'name' => 'col',
            'required' => true,
            'unique' => false,
            'localized' => false,
            'admin' => [
              'isHidden' => false,
              'editor' => ['placement' => 'main', 'width' => 12],
              'fieldType' => 'normal',
            ],
            'readonly' => false,
            'hide' => false,
            'type' => 'string',
            'primaryString' => false,
          ],
        ],
        'softDelete' => false,
        'draftable' => false,
        'sorting' => false,
        'sharable' => true,
        'ownable' => true,
      ],
    ],
  ],
];
"
`;
